src/printer.js

```diff
@@ -3248,6 +3283,118 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+function hasNakedLeftSide(node) {
+  return (
+    node.type === "AssignmentExpression" ||
+    node.type === "BinaryExpression" ||
+    node.type === "LogicalExpression" ||
+    node.type === "ConditionalExpression" ||
+    node.type === "CallExpression" ||
+    node.type === "MemberExpression" ||
+    node.type === "SequenceExpression" ||
+    node.type === "TaggedTemplateExpression"
+  );
+}
+
+function getLeftSide(node) {
+  if (node.expressions) {
+    return node.expressions[0];
+  }
+  return node.left || node.test || node.callee || node.object || node.tag;
+}
+
+function exprNeedsASIProtection(node) {
+  // HACK: node.needsParens is added in `genericPrint()` for the sole purpose
+  // of being used here. It'd be preferable to find a cleaner way to do this.
+  const maybeASIProblem = node.needsParens ||
+    node.type === "ParenthesizedExpression" ||
+    node.type === "TypeCastExpression" ||
+    (node.type === "ArrowFunctionExpression" &&
+      !canPrintParamsWithoutParens(node)) ||
+    node.type === "ArrayExpression" ||
+    node.type === "ArrayPattern" ||
+    (node.type === "UnaryExpression" &&
+      node.prefix &&
+      (node.operator === "+" || node.operator === "-")) ||
+    node.type === "TemplateLiteral" ||
+    node.type === "TemplateElement" ||
+    node.type === "JSXElement" ||
+    node.type === "RegExpLiteral" ||
+    (node.type === "Literal" && node.pattern) ||
+    (node.type === "Literal" && node.regex);
+
+  if (maybeASIProblem) {
+    return true;
+  }
+
+  if (!hasNakedLeftSide(node)) {
+    return false;
+  }
+
+  return exprNeedsASIProtection(getLeftSide(node));
```

Comment:
If you were dealing with path instead, you'd do `path.call(exprNeedsASIProtection, getLeftSideName(path))` or something like that. But I don't know.